common/src/primitives/amount/rpc.rs

```diff
@@ -0,0 +1,169 @@
+/// in both decimal and atom formats.
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct RpcAmountOut {
+    #[serde(flatten)]
+    atoms: Amount,
+
+    #[serde(flatten)]
+    decimal: DecimalAmount,
+}
+
+impl RpcAmountOut {
+    pub const ZERO: Self = Self {
+        atoms: Amount::ZERO,
+        decimal: DecimalAmount::ZERO,
+    };
+
+    fn new_internal(atoms: Amount, decimal: DecimalAmount) -> Self {
+        Self { atoms, decimal }
+    }
+
+    pub fn from_amount(amount: Amount, decimals: u8) -> Self {
+        Self::from_amount_minimal(amount, decimals)
+    }
+
+    /// Construct from amount, keeping all decimal digits
+    pub fn from_amount_full(amount: Amount, decimals: u8) -> Self {
+        let decimal = DecimalAmount::from_amount_full(amount, decimals);
+        Self::new_internal(amount, decimal)
+    }
+
+    /// Construct from amount, keeping the minimal number of decimal places
+    pub fn from_amount_minimal(amount: Amount, decimals: u8) -> Self {
+        let decimal = DecimalAmount::from_amount_minimal(amount, decimals);
+        Self::new_internal(amount, decimal)
+    }
+
+    pub fn amount(&self) -> Amount {
+        self.atoms
+    }
+
+    pub fn decimal(&self) -> DecimalAmount {
+        self.decimal
+    }
+
+    /// Check this is the same number presented in the same way
+    pub fn is_same(&self, other: &Self) -> bool {
+        self.atoms == other.atoms && self.decimal.is_same(&other.decimal)
+    }
+}
+
+impl HasValueHint for RpcAmountOut {
+    const HINT: VH = VH::Object(&[("atoms", &VH::NUMBER_STRING), ("decimal", &VH::DECIMAL_STRING)]);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::rstest;
+    use serde_json::json;
+
+    #[rstest]
+    #[case::zero(0, 0)]
+    #[case(123456, 0)]
+    #[case(123450, 1)]
+    #[case(123450, 2)]
+    #[case(123450, 9)]
+    #[case(4000000000, 3)]
+    #[case(1u128 << 55, 17)]
+    #[case(u128::MAX, 0)]
+    #[case(u128::MAX, 1)]
+    #[case(u128::MAX, 25)]
+    fn rpc_amount_serde(#[case] atoms: u128, #[case] n_decimals: u8) {
+        let atoms = Amount::from_atoms(atoms);
+        let decimal = DecimalAmount::from_amount_minimal(atoms, n_decimals);
+
+        let decimal_str = decimal.to_string();
+        let atoms_str = atoms.into_atoms().to_string();
+
+        assert!(!atoms_str.contains('.'));
+
+        let decimal_in = RpcAmountIn::from_decimal(decimal);
+        let decimal_in_json = json!({ "decimal": decimal_str });
+        assert_eq!(serde_json::to_value(decimal_in).unwrap(), decimal_in_json);
+        assert!(decimal_in.is_same(&serde_json::from_value(decimal_in_json).unwrap()));
+
+        let atoms_in = RpcAmountIn::from_atoms(atoms);
+        let atoms_in_json = json!({ "atoms": atoms_str });
+        assert_eq!(serde_json::to_value(atoms_in).unwrap(), atoms_in_json);
+        assert!(atoms_in.is_same(&serde_json::from_value(atoms_in_json).unwrap()));
+
```

Comment:
I just want to confirm to ensure I understand how this works, even though it's written here: When the user uses RpcAmountIn as a parameter to input amount information, THEY MUST include a json object with being either `{"atoms": "12345"}` or `{"decimal": "123.45"}`. AND, there's no assumptions about `decimal`, as it's stored as a string, and the handler has to decide the number of decimals based on the application (coin vs token). Am I getting this correctly?

Reply:
> I just want to confirm to ensure I understand how this works, even though it's written here: When the user uses RpcAmountIn as a parameter to input amount information, THEY MUST include a json object with being either `{"atoms": "12345"}` or `{"decimal": "123.45"}`. AND, there's no assumptions about `decimal`, as it's stored as a string, and the handler has to decide the number of decimals based on the application (coin vs token). Am I getting this correctly?

Yes, correct. When decimal representation is used, it behaves as `DecimalAmount`. It is stored as a sort of floating point representation holding the integral amount and the number of decimal digits used in the original string representation. In order to convert it to `Amount`, the number of decimals of the target coin/token has to be provided and appropriate conversion is done.